Imbalanced Dataset:

A dataset with an unequal class distribution.

In [2]:
# importing the dependencies
import numpy as np
import pandas as pd

In [3]:
# load the csv data into pandas dataframes
credit_card_data = pd.read_csv("/content/credit_data.csv")

In [4]:
credit_card_data.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [5]:
credit_card_data.tail()


Time         V1         V2        V3        V4        V5  \
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[5 rows x 31 columns]

In [7]:
credit_card_data.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [8]:
# determine the distribution of the two classes
credit_card_data['Class'].value_counts()

Class
0    284315
1       492
Name: count, dtype: int64

This is Highly Imbalanced Dataset.

0 --> Legit Transactions

1 --> Fraudulent Transactions

In [9]:
# seperating the legit and fraud transactions
legit = credit_card_data[credit_card_data.Class == 0]
fraud = credit_card_data[credit_card_data.Class == 1]

In [11]:
print(legit.shape)
print(fraud.shape)

(284315, 31)
(492, 31)


Implementing a technique 'Under-Sampling'.

Building a sample dataset containing similar distribution of legit and fraudulent transactions.

Number of Fraudulent Transactions - 492

Number of Legit Transactions - 284315

In [12]:
# creating a ne df with 492 row of legit transations
legit_sample = legit.sample(n=492)

In [13]:
legit_sample.shape

(492, 31)

concatenate the two dataframes - legit_sample, fraud

In [14]:
new_dataset = pd.concat([legit_sample, fraud], axis=0)

In [15]:
new_dataset.shape

(984, 31)

In [17]:
new_dataset.head()

Time        V1        V2        V3        V4        V5        V6  \
73299  55075.0  1.134524 -1.053062  1.391897 -0.549176 -1.957790 -0.263273   
46698  42891.0 -1.603576  0.349338  2.551085  1.716093 -0.860563  1.160539   
46675  42882.0  0.907100 -0.128511  1.264565  2.502682  0.218227  2.941086   
68043  52804.0  1.103560 -0.205514  0.592762  0.896834 -0.781730 -0.645902   
6502    7789.0 -0.618658  1.155592  1.602665 -0.166635  0.642959 -0.009082   

             V7        V8        V9  ...       V21       V22       V23  \
73299 -1.282454  0.156607 -0.482088  ... -0.056687  0.152723  0.026578   
46698 -1.244092  1.033423  0.741135  ...  0.122145  0.837807 -0.417862   
46675 -1.044683  0.963913  0.183713  ...  0.172061  0.981542 -0.055651   
68043 -0.091964 -0.061968  0.656929  ... -0.271686 -0.797190  0.033894   
6502   1.093564 -1.657159  1.545705  ...  0.187674 -1.084882 -0.052685   

            V24       V25       V26       V27       V28  Amount  Class  
73299  0.576860 -0.125350  1.077748 -0.018894  0.031618   79.99      0  
46698  0.314428 -0.216293 -0.134264  0.231522 -0.034188    1.00      0  
46675 -1.284994  0.321069  0.343970  0.116827 -0.001001    0.00      0  
68043  0.393912  0.284841  0.176723 -0.032662  0.025984   67.99      0  
6502  -0.540372 -0.162015 -0.016300 -0.447311 -0.576416   40.98      0  

[5 rows x 31 columns]

In [18]:
new_dataset.tail()

Time        V1        V2        V3        V4        V5        V6  \
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  ...       V21       V22       V23  \
279863 -0.882850  0.697211 -2.064945  ...  0.778584 -0.319189  0.639419   
280143 -1.413170  0.248525 -1.127396  ...  0.370612  0.028234 -0.145640   
280149 -2.234739  1.210158 -0.652250  ...  0.751826  0.834108  0.190944   
281144 -2.208002  1.058733 -1.632333  ...  0.583276 -0.269209 -0.456108   
281674  0.223050 -0.068384  0.577829  ... -0.164350 -0.295135 -0.072173   

             V24       V25       V26       V27       V28  Amount  Class  
279863 -0.294885  0.537503  0.788395  0.292680  0.147968  390.00      1  
280143 -0.081049  0.521875  0.739467  0.389152  0.186637    0.76      1  
280149  0.032070 -0.739695  0.471111  0.385107  0.194361   77.89      1  
281144 -0.183659 -0.328168  0.606116  0.884876 -0.253700  245.00      1  
281674 -0.450261  0.313267 -0.289617  0.002988 -0.015309   42.53      1  

[5 rows x 31 columns]

In [20]:
new_dataset['Class'].value_counts()

Class
0    492
1    492
Name: count, dtype: int64

Now we have evenly distributed datasets for two classes. When we use this dataset for our prediction this will give a better result.